Add client certificate lookup for Envoy proxy

tsaarni opened the pull request:
This update introduces the ability to lookup X509 client certificates from Envoy's `x-forwarded-client-cert` header (XFCC), using either `Cert` or `Chain` parameters ([doc link](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-client-cert)).

Fixes #33160 

**Motivation**

Built-in support for X509 client certificate lookup from Envoy is a valuable addition to Keycloak, as Envoy is widely adopted, particularly within the Kubernetes ecosystem. It serves as the core ingress proxy for projects like Istio, Contour, and others.

**Testing** 

Since there don't appear to be existing tests for X509 client certificate lookups, ~this PR does not include automated tests~. However, I have conducted manual testing with both Istio and Contour, see [notes here](https://gist.github.com/tsaarni/37911f2fe569c060b4debb66ed0f7533).

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

tsaarni committed:
Add client certificate lookup for Envoy proxy

This update introduces the ability to lookup X509 client certificates from
Envoy's x-forwarded-client-cert header, using either the Cert or Chain
parameters.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

ahus1 commented:
@tsaarni - thank you for this PR. I'll write here a similar comment like in #31778

It would be great to have an out-of-the-box support for Envoy. At the same time it would difficult and probably out-of-scope for the Keycloak project test this with every release.

AFAIK to have a full integration test, one would need to spin up some Kubernetes resources. For a more modular test, some JUnit tests would be sufficient, still those are missing in this PR.

This PR shows that Keycloak has the right SPIs aka extension points for this. If Keycloak's core functionality needs to be enhanced, this can of course be done in a PR.

So I would consider this a great candidate for a Keycloak community extension that can then be maintained by those familiar with Envoy. We list all extensions on Keycloak's homepage, and we would be happy to add it once it exists: https://www.keycloak.org/extensions.html

cc: @vmuzikar / @stianst

tsaarni committed:
added junit test cases for envoy x509 lookup

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
Thank you, @ahus1, for your response!

I've added JUnit tests in this PR. I don’t believe Envoy is likely to modify the header format, so a full end-to-end test may not be necessary.

I want to confirm—did I understand correctly that built-in Envoy support won't be part of the Keycloak codebase, meaning this PR won’t be merged?

Istio has gained a lot of popularity, and there's even the [OpenShift Service Mesh](https://www.redhat.com/en/technologies/cloud-computing/openshift/what-is-openshift-service-mesh). ;)

ahus1 commented:
> I want to confirm—did I understand correctly that built-in Envoy support won't be part of the Keycloak code base, meaning this PR won’t be merged?

Yes. At least not for now. If it turns out to become a popular extension, we might reconsider it.

Having this in an extension, this would allow you to include more examples like you showed in the README you linked on how to use it. To maintain those examples, this would require knowledge of both Envoy (and probably other tools like Istio and Contour) which we don't have in the core maintainers of Keycloak as of today. Maintaining would include testing it when new releases of either Keycloak or Envoy/Istio/Contour become available.

The community that would form around that extension could also provide documentation on how this works with different versions of Istio/Contour/Envoy/Keycloak, and would test them accordingly. 

From the Keycloak side of things, when there are multiple extensions using this, we could consider making this a public SPI (today it is in keycloak-services), to make sure it stays stable across Keycloak releases.

vmuzikar commented:
@tsaarni Thank you for the PR.

I see it similarly as @ahus1. Having this on the official extensions list works for me and I agree the main concern about having this in the Keycloak code base is maintainability. That said, there are other similar Client Certificate providers that are part of the code base – [HA](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/x509/HaProxySslClientCertificateLookup.java), [Apache](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/x509/ApacheProxySslClientCertificateLookup.java), [Nginx](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/x509/NginxProxyTrustedClientCertificateLookup.java). But they are not really covered by integration tests so this is not ideal and we probably should not another one.

tsaarni commented:
I've created a standalone project from this and put it here https://github.com/Nordix/keycloak-client-cert-lookup-for-envoy. Loading it in Keycloak will produce a warning "...is implementing the internal SPI x509cert-lookup. This SPI is internal and may change without notice" but otherwise it works.

tsaarni commented:
> Create a new issue advocating to make SPI public to ensure it is stable and Keycloak won't log the warning. Post the link to that new issue here, so I can follow-up on it.

Thanks @ahus1. I've submitted #33818.

tsaarni committed:
Run spotless

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni committed:
Added support for proxy-trusted-addresses

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
I've now resolved conflicts and added call to `isProxyTrusted()` from #35861, which was introduced after my last update to this PR.

I understand this is outside the scope of this PR: as far as I understand, IP allowlist is not usable in Kubernetes since proxy does not have stable address. In my external Envoy plugin, I added authorization based on the subject name from the proxy's client certificate from TLS connection between the proxy and Keycloak, ensuring that only authorized peers can provide the certificate in a header. I believe this is the correct approach. For more details, see the diagram here: https://github.com/Nordix/keycloak-client-cert-lookup-for-envoy/blob/main/docs/security-and-client-cert-forwarding.md. Could we do something like that in Keycloak?

tsaarni committed:
Convert exception to SecurityException for invalid headers

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

shawkins commented on docs/guides/server/reverseproxy.adoc:
```diff
@@ -203,6 +203,9 @@ The server supports some of the most commons TLS termination proxies such as:
 |Proxies that are compliant to link:https://datatracker.ietf.org/doc/rfc9440/[RFC 9440]
+
+|Envoy
+|envoy
```
```suggestion
|envoy
|Envoy
```

shawkins reviewed (commented):
This looks good. Just a couple of thoughts.

The common options imply that the header and chain length options apply to envoy as well https://github.com/keycloak/keycloak/blob/c8428c040a896696d9d0e99b37df509e59c6c7a4/docs/guides/server/reverseproxy.adoc?plain=1#L227

For most of the providers I'm unsure, other than performance, why the chain length is limited. So if it makes sense to you to return the full chain, just update the doc to exclude envoy.

tsaarni committed:
document update

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

shawkins reviewed (approved):
LGTM, thanks @tsaarni

vmuzikar reviewed (approved):
LGTM, thanks.
